runtime/near-evm-runner/src/pricer.rs

```diff
@@ -0,0 +1,430 @@
+#[derive(Debug)]
+pub struct Linear {
+    pub base: u64,
+    pub word: u64,
+}
+
+/// A special pricing model for modular exponentiation.
+#[derive(Debug)]
+pub struct ModexpPricer {
+    pub divisor: u64,
+}
+
+impl Pricer for Linear {
+    fn cost(&self, input: &[u8]) -> U256 {
+        U256::from(self.base) + U256::from(self.word) * U256::from((input.len() + 31) / 32)
+    }
+}
+
+/// alt_bn128 pairing price
+#[derive(Debug, Copy, Clone)]
+pub struct AltBn128PairingPrice {
+    pub base: u64,
+    pub pair: u64,
+}
+
+/// alt_bn128_pairing pricing model. This computes a price using a base cost and a cost per pair.
+#[derive(Debug)]
+pub struct AltBn128PairingPricer {
+    pub price: AltBn128PairingPrice,
+}
+
+/// Pricing for constant alt_bn128 operations (ECADD and ECMUL)
+#[derive(Debug, Copy, Clone)]
+pub struct AltBn128ConstOperations {
+    /// Fixed price.
+    pub price: u64,
+}
+
+impl Pricer for AltBn128ConstOperations {
+    fn cost(&self, _input: &[u8]) -> U256 {
+        self.price.into()
+    }
+}
+
+impl Pricer for AltBn128PairingPricer {
+    fn cost(&self, input: &[u8]) -> U256 {
+        U256::from(self.price.base) + U256::from(self.price.pair) * U256::from(input.len() / 192)
+    }
+}
+
+impl Pricer for ModexpPricer {
+    fn cost(&self, input: &[u8]) -> U256 {
+        let mut reader = input.chain(io::repeat(0));
+        let mut buf = [0; 32];
+
+        // read lengths as U256 here for accurate gas calculation.
+        let mut read_len = || {
+            reader
+                .read_exact(&mut buf[..])
+                .expect("reading from zero-extended memory cannot fail; qed");
+            U256::from_big_endian(&buf[..])
+        };
+        let base_len = read_len();
+        let exp_len = read_len();
+        let mod_len = read_len();
+
+        if mod_len.is_zero() && base_len.is_zero() {
+            return U256::zero();
+        }
+
+        let max_len = U256::from(u32::max_value() / 2);
+        if base_len > max_len || mod_len > max_len || exp_len > max_len {
+            return U256::max_value();
+        }
+        let (base_len, exp_len, mod_len) =
+            (base_len.low_u64(), exp_len.low_u64(), mod_len.low_u64());
+
+        let m = max(mod_len, base_len);
+        // read fist 32-byte word of the exponent.
+        let exp_low = if base_len + 96 >= input.len() as u64 {
+            U256::zero()
+        } else {
+            buf.iter_mut().for_each(|b| *b = 0);
+            let mut reader = input[(96 + base_len as usize)..].chain(io::repeat(0));
+            let len = min(exp_len, 32) as usize;
+            reader
+                .read_exact(&mut buf[(32 - len)..])
+                .expect("reading from zero-extended memory cannot fail; qed");
+            U256::from_big_endian(&buf[..])
+        };
+
+        let adjusted_exp_len = Self::adjusted_exp_len(exp_len, exp_low);
+
+        let (gas, overflow) = Self::mult_complexity(m).overflowing_mul(max(adjusted_exp_len, 1));
```

Comment:
FYI. No action needed.
This code is terrible and hard to read. A lot of overflow checks are just coming from the fact that some variable was limited in length. 
Why this is limited to `u64`? Is Ethereum maximum gas value limited to `u64`? I assume yes, so it's probably okay.

Reply:
didn't find anywhere documented maximum ethereum gas, but i guess no one has enough ether to prepay more than u64::max amount of gas